File: 2주차/김용상/정수 삼각형.java
import java.util.*;
import java.lang.*;
class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int n = triangle.length;
        // 아래서부터 부분 최적값 구하기
        List<Integer>[]list = new ArrayList[n-1];
        for (int i=0;i<n-1;i++){
            list[i] = new ArrayList<>();
        }
        
        for (int i=n-2;i>-1;i--){
            for (int j=0;j<triangle[i].length;j++){
                int mx = Math.max(triangle[i+1][j],triangle[i+1][j+1]);
                triangle[i][j] +=mx;
            
            }
        }
        return triangle[0][0];
    }
}

File: 2주차/이라은/정수 삼각형.py
def solution(triangle):    
    dp = []
    for i in range(len(triangle)):
        row = []
        for j in range(len(triangle[i])):
            row.append(0)
        dp.append(row)
    dp[0][0]=triangle[0][0]
    dp[1][0]=triangle[1][0] + triangle[0][0]
    dp[1][1]=triangle[1][1] + triangle[0][0]
    
    for i in range(2, len(triangle)):
        length = len(triangle[i])
        for j in range(length):
            if j == 0:
                dp[i][0] += triangle[i][0] + dp[i-1][0]
            elif j == length-1:
                dp[i][length-1] += triangle[i][-1] + dp[i-1][-1]
            else: 
                dp[i][j] += max((triangle[i][j] + dp[i-1][j-1]), (dp[i-1][j] + triangle[i][j]))

    return max(dp[-1])
